diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/SingleConstructorRule.kt

```diff
@@ -0,0 +1,151 @@
+     * - Inside the single secondary constructor find all assignments.
+     * - Some of assigned values will have `this` qualifier, they are definitely class properties.
+     * - For other assigned variables that are not declared in the same scope we check if they are properties.
+     * - Create primary constructor moving all properties that we collected.
+     * - Create init block with other statements from the secondary constructor.
+     * - Finally, remove the secondary constructor.
+     */
+    @Suppress("UnsafeCallOnNullableType")
+    private fun convertConstructorToPrimary(classNode: ASTNode, secondaryCtor: ASTNode) {
+        val (assignments, otherStatements) = (secondaryCtor.psi as KtSecondaryConstructor)
+            .bodyBlockExpression
+            ?.statements
+            ?.partition { (it as? KtBinaryExpression)?.asAssignment() != null }
+            ?: emptyList<KtExpression>() to emptyList()
+
+        val (qualifiedAssignedProperties, assignedProperties) = assignments
+            .map {
+                // non-null assert is safe because of predicate in partitioning
+                (it as KtBinaryExpression).asAssignment()!!.left!!
+            }
+            .filter {
+                // todo use secondaryCtor.findAllVariablesWithAssignments()
+                (it as? KtDotQualifiedExpression)?.run {
+                    receiverExpression is KtThisExpression && selectorExpression is KtNameReferenceExpression
+                } ?: false ||
+                        it is KtNameReferenceExpression
+            }
+            .partition { it is KtDotQualifiedExpression }
+            .let { (qualified, simple) ->
+                qualified.map { (it as KtDotQualifiedExpression).selectorExpression as KtNameReferenceExpression } to
+                        simple.map { it as KtNameReferenceExpression }
+            }
+
+        val assignedClassProperties = qualifiedAssignedProperties + assignedProperties
+            .run {
+                val localProperties = secondaryCtor.findAllNodesWithSpecificType(ElementType.PROPERTY)
+                filterNot { reference ->
+                    // check for shadowing
+                    localProperties.any {
+                        reference.node.isGoingAfter(it) && (it.psi as KtProperty).name == reference.name
+                    }
+                }
+            }
+
+        // find properties declarations
+        val declarationsAssignedInCtor = assignedClassProperties
+            .mapNotNull { reference ->
+                (classNode.psi as KtClass).getProperties()
+                    .firstOrNull { it.nameIdentifier?.text == reference.getReferencedName() }
+            }
+            .distinct()
+
+        classNode.convertSecondaryConstructorToPrimary(secondaryCtor, declarationsAssignedInCtor, otherStatements)
+    }
+
+    @Suppress("UnsafeCallOnNullableType", "NestedBlockDepth")
+    private fun ASTNode.convertSecondaryConstructorToPrimary(secondaryCtor: ASTNode,
```

Comment:
I cannot find where you move modifiers like `private`

Reply:
Totally forgot about it. Thanks, fixed.